library/firewall_lib.py

```diff
@@ -299,6 +301,29 @@ def main():
             "icmp_block_inversion, target or zone needs to be set"
         )
 
+    if state == "present" or state == "absent":
+        if zone is None:
+            module.fail_json(
+                msg="Zone must be set in order to use present or absent state"
+            )
+        elif (
+            masquerade is not None
+            and icmp_block_inversion is not None
+            and target is not None
+            and len(service)
+            + len(port)
+            + len(source_port)
+            + len(forward_port)
+            + len(rich_rule)
+            + len(source)
+            + len(interface)
+            + len(icmp_block)
+            > 0
+        ):
+            module.fail_json(
+                msg="The states present and absent can only be used in zone level operations (i.e. when no other parameters but zone and state are set)."
```

Comment:
In my opinion `present` or `enabled` should always ensure that a zone exists. If the zone does not exist, it should simply create a zone with the given settings.
"absent" with zone parameters should ensure the absence of the given settings within the zone. Without zone parameters it should ensure the absence of the zone. To have a better differentiation about the presence and absence of items in lists it might be good to have an additional parameter. In ansible-freeipa for example we are using the `action` parameter for this. (Example: https://github.com/freeipa/ansible-freeipa/blob/master/README-hostgroup.md)